### PR TITLE
Implementation of the open feature SDK in the java tracer

### PR DESCRIPTION
**Pull request by manuel-alvarez-alvarez**
# What Does This Do
Implements the integration of the OpenFeature SDK within the Java tracer. This includes a new umbrella module called `:products:feature-flagging` that includes:
- `:agent` feature flagging system to be initialized by the agent during startup.
- `:api` implementation of the [OpenFeature Java provider](https://openfeature.dev/docs/reference/sdks/server/java/)
- `:lib` for feature flag configuration and exposure submission through the EVP proxy.
- `:bootstrap` bridge exposing the `FeatureFlagGateway` API between the internal `:agent` (datadog classloader) and the `:api` (application classloader)

# Motivation
This adds OpenFeature SDK support to the tracer so feature flag evaluations can be managed and reported consistently through the Datadog feature flagging system.

# Additional Notes
Smoke tests are provided as a separate [PR](https://github.com/DataDog/dd-trace-java/pull/9961) 

# Contributor Checklist

- Format the title [according the contribution guidelines](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#title-format)
- Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any useful labels](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#labels)
- Don't use `close`, `fix` or any [linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue.  
  Use `solves` instead, and assign the PR [milestone](https://github.com/DataDog/dd-trace-java/milestones) to the issue
- Update the [CODEOWNERS](https://github.com/DataDog/dd-trace-java/blob/master/.github/CODEOWNERS) file on source file addition, move, or deletion
- Update the [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) in case of new configuration flag or behavior

Jira ticket: [PROJ-IDENT]

[RFC](https://docs.google.com/document/d/1M4delPpJrBOP437aj8QZRMvDc8JNV260oLMuLZWTwZ0/edit?tab=t.0#heading=h.9omoxl2muhqi)

### CODE REVIEW

File: gradle/libs.versions.toml [hunk 1]
```diff
@@ -104,6 +105,8 @@ jmc-flightrecorder = { module = "org.openjdk.jmc:flightrecorder", version.ref =
 
 jackson-databind = {module = "com.fasterxml.jackson.core:jackson-databind", version.ref = "jackson"}
 
+openfeature-sdk = {module = "dev.openfeature:sdk", version.ref = "openfeature"}
```
Comment (PerfectSlayer) [inline gradle/libs.versions.toml]: **🎯 suggestion:** ‏When dependencies are product related - rather than core / framework - I would rather not adding them to the common lib catalog. You can still created dedicated lib catalog if you want to have one, or import the library as usual from the build file.

Comment (manuel-alvarez-alvarez) [reply gradle/libs.versions.toml]: Removed from the dedicated catalog :wink:

**Comment by manuel-alvarez-alvarez** [comment]
> **🎯 suggestion:** ‏Should we start having dedicated API module for the other products? Because this is does not feel like "tracing" and I would rather not having sub-modules and third party dependencies from our public tracing API.
> 
> So what about starting a `:products:open-features` for your public API? Because the direction for 2.x version will be to split all products from tracing and have them own their separate API which can have 3rd party dependency and language version requirement like yours.

Hello @PerfectSlayer, thanks for the input! I did take your suggestion into account and also removed the draft status from the PR, so feel free to continue doing the review. We should also discuss the code owners from the new module (including members of the FFE team)

File: .github/CODEOWNERS [hunk 2]
```diff
@@ -139,3 +139,10 @@ dd-trace-core/src/test/groovy/datadog/trace/llmobs/  @DataDog/ml-observability
 /internal-api/src/test/groovy/datadog/trace/api/rum/ @DataDog/rum
 /telemetry/src/main/java/datadog/telemetry/rum/      @DataDog/rum
 /telemetry/src/test/groovy/datadog/telemetry/rum/    @DataDog/rum
+
+# @DataDog/feature-flagging-and-experimentation-sdk
+/dd-java-agent/agent-featureflag/                           @DataDog/feature-flagging-and-experimentation-sdk
```
Comment (leoromanovsky) [inline .github/CODEOWNERS]: 👍

File: products/feature-flagging/lib/src/main/java/com/datadog/featureflag/ExposureWriterImpl.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import org.jctools.queues.MpscBlockingConsumerArrayQueue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ExposureWriterImpl implements ExposureWriter {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ExposureWriterImpl.class);
+  private static final int DEFAULT_CAPACITY = 1 << 16; // 65536 elements
```
Comment (AlexeyKuznetsov-DD) [inline products/feature-flagging/lib/src/main/java/com/datadog/featureflag/ExposureWriterImpl.java:30]: Just curious, why not to use 65536 directly?

File: dd-java-agent/agent-featureflag/src/main/java/com/datadog/featureflag/ExposureWriterImpl.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+      final HttpUrl agentUrl,
+      final Config config) {
+    this.queue = new MpscBlockingConsumerArrayQueue<>(capacity);
+    final Headers headers = Headers.of(EVP_SUBDOMAIN_HEADER_NAME, EVP_SUBDOMAIN_HEADER_VALUE);
+    final HttpUrl url =
+        HttpUrl.get(
+            agentUrl.toString()
+                + DDAgentFeaturesDiscovery.V2_EVP_PROXY_ENDPOINT
+                + EXPOSURES_API_PATH);
+    final Map<String, String> context = new HashMap<>();
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/agent-featureflag/src/main/java/com/datadog/featureflag/ExposureWriterImpl.java]: Does it make sense to specify size of `context` as `2`?

File: dd-java-agent/agent-feature-flagging/build.gradle [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  id 'com.gradleup.shadow'
+}
+
+apply from: "$rootDir/gradle/java.gradle"
+apply from: "$rootDir/gradle/version.gradle"
+
+java {
+  sourceCompatibility = JavaVersion.VERSION_1_8
+  targetCompatibility = JavaVersion.VERSION_1_8
+}
```
Comment (AlexeyKuznetsov-DD) [inline dd-java-agent/agent-feature-flagging/build.gradle]: Not 100% sure, but with latest changes in build scripts we have this as default.
@bric3 Could you confirm on this?

File: dd-smoke-tests/featureflag/src/main/java/datadog/smoketest/springboot/SpringbootApplication.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+import java.lang.management.ManagementFactory;
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
+public class SpringbootApplication {
+
+  public static void main(final String[] args) {
+    SpringApplication.run(SpringbootApplication.class, args);
+    System.out.println("Started in " + ManagementFactory.getRuntimeMXBean().getUptime() + "ms");
```
Comment (AlexeyKuznetsov-DD) [inline dd-smoke-tests/featureflag/src/main/java/datadog/smoketest/springboot/SpringbootApplication.java]: I believe that Spring Boot will report the same out of the box?

File: internal-api/src/main/java/datadog/trace/api/featureflag/exposure/Allocation.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(key);
+  }
+
+  @Override
+  public String toString() {
+    return "Allocation{" + "key='" + key + '\'' + '}';
```
Comment (AlexeyKuznetsov-DD) [inline internal-api/src/main/java/datadog/trace/api/featureflag/exposure/Allocation.java]:
```suggestion
    return "Allocation{" + "key='" + key + "'}";
```

File: internal-api/src/main/java/datadog/trace/api/featureflag/exposure/Flag.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(key);
+  }
+
+  @Override
+  public String toString() {
+    return "Flag{" + "key='" + key + '\'' + '}';
```
Comment (AlexeyKuznetsov-DD) [inline internal-api/src/main/java/datadog/trace/api/featureflag/exposure/Flag.java]:
```suggestion
    return "Flag{" + "key='" + key + "'}";
```

File: internal-api/src/main/java/datadog/trace/api/featureflag/exposure/Subject.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(id, attributes);
+  }
+
+  @Override
+  public String toString() {
+    return "Subject{" + "id='" + id + ", attributes=" + attributes + '\'' + '}';
```
Comment (AlexeyKuznetsov-DD) [inline internal-api/src/main/java/datadog/trace/api/featureflag/exposure/Subject.java]:
```suggestion
    return "Subject{" + "id='" + id + ", attributes=" + attributes + "'}";
```

File: internal-api/src/main/java/datadog/trace/api/featureflag/exposure/Variant.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(key);
+  }
+
+  @Override
+  public String toString() {
+    return "Variant{" + "key='" + key + '\'' + '}';
```
Comment (AlexeyKuznetsov-DD) [inline internal-api/src/main/java/datadog/trace/api/featureflag/exposure/Variant.java]:
```suggestion
    return "Variant{" + "key='" + key +"'}";
```

File: products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java [hunk 11]
```diff
@@ -0,0 +1,518 @@
...
+    }
+
+    return String.valueOf(a).equals(String.valueOf(b));
+  }
+
+  private static boolean compareNumber(
+      final Object attributeValue, final Object conditionValue, NumberComparator comparator) {
+    final double a = mapValue(Double.class, attributeValue);
+    final double b = mapValue(Double.class, conditionValue);
+    return comparator.compare(a, b);
```
Comment (AlexeyKuznetsov-DD) [inline products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java:290]: not 100% but comparing `doubles` can be not stable?

File: products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java [hunk 12]
```diff
@@ -0,0 +1,518 @@
...
+      try {
+        final SimpleDateFormat sdf = new SimpleDateFormat(format);
+        sdf.setTimeZone(TimeZone.getTimeZone("UTC"));
+        return sdf.parse(dateString);
+      } catch (ParseException e) {
+        // ignore it
+      }
+    }
+    return null;
+  }
```
Comment (AlexeyKuznetsov-DD) [inline products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java]: Can this be refactored to use `DateTimeFormatter` ?

File: products/feature-flagging/lib/src/main/java/com/datadog/featureflag/ExposureWriterImpl.java (hunk 3 above)
Comment (manuel-alvarez-alvarez) [reply products/feature-flagging/lib/src/main/java/com/datadog/featureflag/ExposureWriterImpl.java:30]: The value must be a power of two, as this makes the intent more explicit.

File: dd-java-agent/agent-featureflag/src/main/java/com/datadog/featureflag/ExposureWriterImpl.java (hunk 4 above)
Comment (manuel-alvarez-alvarez) [reply dd-java-agent/agent-featureflag/src/main/java/com/datadog/featureflag/ExposureWriterImpl.java]: Yep, I'll set 4 as the map can contain 2 or 3 elements.

File: dd-smoke-tests/featureflag/src/main/java/datadog/smoketest/springboot/SpringbootApplication.java (hunk 6 above)
Comment (manuel-alvarez-alvarez) [reply dd-smoke-tests/featureflag/src/main/java/datadog/smoketest/springboot/SpringbootApplication.java]: Removed

File: internal-api/src/main/java/datadog/trace/api/featureflag/exposure/Allocation.java (hunk 7 above)
Comment (manuel-alvarez-alvarez) [reply internal-api/src/main/java/datadog/trace/api/featureflag/exposure/Allocation.java]: I did remove all this `toString` methods, as they were only useful for debugging the tests initially.

File: products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java (hunk 12 above)
Comment (manuel-alvarez-alvarez) [reply products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java]: Yep, just done it, thanks for the suggestion!

File: products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java (hunk 11 above)
Comment (manuel-alvarez-alvarez) [reply products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java:290]: As far as I know, comparisons between double values are stable and deterministic as long as no arithmetic operations are involved. The tricky cases arise when performing calculations like `0.1 + 0.2 == 0.3`, where rounding errors can lead to unexpected results.

File: internal-api/src/main/java/datadog/trace/api/featureflag/exposure/ExposureEvent.java [hunk 13]
```diff
@@ -0,0 +1,22 @@
+package datadog.trace.api.featureflag.exposure;
+
+public class ExposureEvent {
+  public final long timestamp;
```
Comment (leoromanovsky) [inline internal-api/src/main/java/datadog/trace/api/featureflag/exposure/ExposureEvent.java:5]: In Java is this the canonical way to represent or name milliseconds since epoch? Just mentioning that clarity is useful here because we had this come up in ruby development: https://dd.slack.com/archives/C0959U2JP2N/p1763049792367839?thread_ts=1763044781.435079&cid=C0959U2JP2N

File: products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/Provider.java [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+import dev.openfeature.sdk.exceptions.ProviderNotReadyError;
+import java.lang.reflect.Constructor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class Provider extends EventProvider implements Metadata {
+
+  static final String METADATA = "datadog-openfeature-provider";
+  private static final String EVALUATOR_IMPL = "datadog.trace.api.openfeature.DDEvaluator";
+  private static final Options DEFAULT_OPTIONS = new Options().initTimeout(30, SECONDS);
```
Comment (leoromanovsky) [inline products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/Provider.java:24]: Thanks for keeping this consistent across SDKs

File: products/feature-flagging/api/src/test/java/datadog/trace/api/openfeature/ProviderTest.java [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+    final Client client = api.getClient();
+    client.on(ProviderEvent.PROVIDER_READY, readyEvent);
+
+    // we time out after 10 millis without receiving the initial config
+    assertThrows(
+        ProviderNotReadyError.class,
+        () -> api.setProviderAndWait(new Provider(new Options().initTimeout(10, MILLISECONDS))));
+
+    // ready has not yet been called
+    verify(readyEvent, times(0)).accept(any());
```
Comment (leoromanovsky) [inline products/feature-flagging/api/src/test/java/datadog/trace/api/openfeature/ProviderTest.java:104]: 👍

File: products/feature-flagging/api/src/test/java/datadog/trace/api/openfeature/DDEvaluatorTest.java [hunk 16]
```diff
@@ -0,0 +1,1223 @@
...
+    final DDEvaluator evaluator = new DDEvaluator(mock(Runnable.class));
+    evaluator.accept(mock(ServerConfiguration.class));
+    final ProviderEvaluation<?> details = evaluator.evaluate(Integer.class, "test", 23, null);
+    assertThat(details.getValue(), equalTo(23));
+    assertThat(details.getReason(), equalTo(ERROR.name()));
+    assertThat(details.getErrorCode(), equalTo(ErrorCode.INVALID_CONTEXT));
+  }
+
+  @Test
+  public void testNoAllocations() {
```
Comment (leoromanovsky) [inline products/feature-flagging/api/src/test/java/datadog/trace/api/openfeature/DDEvaluatorTest.java:159]: Just a note for the future: The other SDKs copied the json test files locally then ran a "unit test" level version of the system tests. However now we have copies of the json files all over; we will want to add to these files if new edge cases come up. As a follow up to our project I am going to try to introduce the files from `system-tests` as git submodules into the individual trace libraries so we can keep a canonical set of files and support local updating on the tracers' control.

File: dd-java-agent/agent-feature-flagging/src/main/java/com/datadog/featureflag/LRUExposureCache.java [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+      return true;
+    }
+    final Value newValue = new Value(event);
+    if (!newValue.equals(oldValue)) {
+      cache.remove(key); // ensure LRU semantics
+      cache.put(key, newValue);
+      return true;
+    }
+    return false;
+  }
```
Comment (leoromanovsky) [inline dd-java-agent/agent-feature-flagging/src/main/java/com/datadog/featureflag/LRUExposureCache.java]: This function is not thread safe which could lead to an incorrect cache state. Consider synchronizing access to `cache`

File: dd-java-agent/agent-feature-flagging/src/main/java/com/datadog/featureflag/LRUExposureCache.java [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+    private FIFOCache(final int capacity) {
+      this.capacity = capacity;
+    }
+
+    @Override
+    protected boolean removeEldestEntry(final Map.Entry<K, V> eldest) {
+      return size() > capacity;
+    }
+  }
+}
```
Comment (leoromanovsky) [inline dd-java-agent/agent-feature-flagging/src/main/java/com/datadog/featureflag/LRUExposureCache.java]: My reading of `LinkedHashMap` (https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html) is it will behave closer to FIFO behavior: ` This linked list defines the iteration ordering, which is normally the order in which keys were inserted into the map (insertion-order). ` 

You might be interested in the `accessOrder` parameter:

```
accessOrder - the ordering mode - true for access-order, false for insertion-order
```

Could you please write a test for this class with a small capacity to demonstrate the desired behavior.

File: products/feature-flagging/lib/src/test/groovy/com/datadog/featureflag/LRUExposureCacheTest.groovy [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+    final event4 = createEvent('flag4', 'subject4', 'variant4', 'allocation4')
+    final key1 = new ExposureCache.Key(event1)
+    final key2 = new ExposureCache.Key(event2)
+    final key4 = new ExposureCache.Key(event4)
+
+    when:
+    cache.add(event1)
+    cache.add(event2)
+    cache.add(event3)
+    cache.add(event1Updated) // Updates event1, moves to most recent
```
Comment (leoromanovsky) [inline products/feature-flagging/lib/src/test/groovy/com/datadog/featureflag/LRUExposureCacheTest.groovy:191]: To have the highest cache hit rate, we need to keep very highly evaluating subjects in the cache as much as possible. Almost always the variations will not be changing, we'll just be accessing the cache to see if a key is present. Imagine a power user on an application doing 100x the actions of a regular user.

A more common case that's missing in the test is `cache.add(same event as already cache)`.

This is a test I am thinking of:

```
void 'test duplicate exposure keeps subject hot in LRU order'() {
  given:
  final cache = new LRUExposureCache(3)
  final event1 = createEvent('flag1', 'subject1', 'variant1', 'allocation1')
  final event2 = createEvent('flag2', 'subject2', 'variant2', 'allocation2')
  final event3 = createEvent('flag3', 'subject3', 'variant3', 'allocation3')
  // same key + same details as event1: will go through the "duplicate" path
  final event1Duplicate = createEvent('flag1', 'subject1', 'variant1', 'allocation1')
  final event4 = createEvent('flag4', 'subject4', 'variant4', 'allocation4')

  final key1 = new ExposureCache.Key(event1)
  final key2 = new ExposureCache.Key(event2)
  final key4 = new ExposureCache.Key(event4)

  when:
  // Fill cache
  def added1 = cache.add(event1)
  def added2 = cache.add(event2)
  def added3 = cache.add(event3)

  // Duplicate exposure for subject1: should *not* change size, but *should* bump recency
  def duplicateAdded = cache.add(event1Duplicate)

  // Now push over capacity: the least recently used *non-hot* entry (event2) should be evicted
  def added4 = cache.add(event4)

  then:
  added1
  added2
  added3
  !duplicateAdded     // dedup correctly
  added4

  cache.size() == 3

  cache.get(key1) != null          // hot subject1 should still be present
  cache.get(key2) == null          // subject2 should be evicted
  cache.get(key4) != null          // newest subject4 should be present

  cache.get(key1).variant == 'variant1'
  cache.get(key1).allocation == 'allocation1'
}
```

File: products/feature-flagging/api/src/test/java/datadog/trace/api/openfeature/DDEvaluatorTest.java (hunk 16 above)
Comment (manuel-alvarez-alvarez) [reply products/feature-flagging/api/src/test/java/datadog/trace/api/openfeature/DDEvaluatorTest.java:159]: Yes — you can find them in the smoke test PR: https://github.com/DataDog/dd-trace-java/pull/9961

We decided to separate them to make the review process more manageable for everyone 😊

File: dd-java-agent/agent-feature-flagging/src/main/java/com/datadog/featureflag/LRUExposureCache.java (hunk 17 above)
Comment (manuel-alvarez-alvarez) [reply dd-java-agent/agent-feature-flagging/src/main/java/com/datadog/featureflag/LRUExposureCache.java]: It's not intended to be thread safe, only a single thread access the LRU cache. I added a comment to the class to make it clear.

File: products/feature-flagging/lib/src/test/groovy/com/datadog/featureflag/LRUExposureCacheTest.groovy (hunk 19 above)
Comment (manuel-alvarez-alvarez) [reply products/feature-flagging/lib/src/test/groovy/com/datadog/featureflag/LRUExposureCacheTest.groovy:191]: Thanks! I had some doubts while implementing, just updated the code to keep recency up to date.

File: internal-api/src/main/java/datadog/trace/api/featureflag/exposure/ExposureEvent.java (hunk 13 above)
Comment (manuel-alvarez-alvarez) [reply internal-api/src/main/java/datadog/trace/api/featureflag/exposure/ExposureEvent.java:5]: Yes, in Java we have: https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#currentTimeMillis-- let me add a comment to make it clear.

File: .github/CODEOWNERS [hunk 20]
```diff
@@ -139,3 +139,10 @@ dd-trace-core/src/test/groovy/datadog/trace/llmobs/  @DataDog/ml-observability
 /internal-api/src/test/groovy/datadog/trace/api/rum/ @DataDog/rum
 /telemetry/src/main/java/datadog/telemetry/rum/      @DataDog/rum
 /telemetry/src/test/groovy/datadog/telemetry/rum/    @DataDog/rum
+
+
+# @DataDog/feature-flagging-and-experimentation-sdk
```
Comment (PerfectSlayer) [inline .github/CODEOWNERS]: **🔍 nitpick:** ‏Can you keep the list ordered alphabetically?

File: internal-api/src/main/java/datadog/trace/api/intake/Intake.java [hunk 21]
```diff
@@ -15,13 +15,18 @@ public enum Intake {
...
-      Config::getCiVisibilityIntakeAgentlessUrl);
+      Config::getCiVisibilityIntakeAgentlessUrl),
+  EVENT_PLATFORM("event-platform-intake", "v2");
 
   public final String urlPrefix;
   public final String version;
   public final Function<Config, Boolean> agentlessModeEnabled;
   public final Function<Config, String> customUrl;
 
+  Intake(String urlPrefix, String version) {
```
Comment (PerfectSlayer) [inline internal-api/src/main/java/datadog/trace/api/intake/Intake.java:26]: **🎯 suggestion:** ‏I would use a static factory with a dedicate name (like `forBackend()` but you can definitely find something better) that will explicit the default constructor values.

File: products/feature-flagging/agent/build.gradle.kts [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+    val deps = project.extra["deps"] as Map<*, *>
+    val excludeShared = deps["excludeShared"] as groovy.lang.Closure<*>
+    excludeShared.delegate = this
+    excludeShared.call()
+  }
+}
+
+tasks.named<Jar>("jar") {
+  archiveClassifier.set("unbundled")
+}
```
Comment (PerfectSlayer) [inline products/feature-flagging/agent/build.gradle.kts]: **❔ question:** ‏Dumb question but what is the `jar` / `shadowJar` tasks are for?

File: products/feature-flagging/api/build.gradle.kts [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+
+java {
+  toolchain {
+    languageVersion = JavaLanguageVersion.of(11)
+  }
+}
+
+dependencies {
+  api(libs.slf4j)
+  api("dev.openfeature:sdk:1.18.2")
```
Comment (PerfectSlayer) [inline products/feature-flagging/api/build.gradle.kts:42]: **❔ question:** Related to your comment:

> No, no, the dev.openfeature:sdk will be a transitive dependency of the provider (it won't be included in Datadog's tracer)

Is the dependency trimmed by the `shadowJar` task from the openfeature agent module then?

File: internal-api/src/main/java/datadog/trace/api/featureflag/exposure/Allocation.java [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+
+public class Allocation {
+  public final String key;
+
+  public Allocation(final String key) {
+    this.key = key;
+  }
+
+  @Override
+  public boolean equals(final Object o) {
```
Comment (PerfectSlayer) [inline internal-api/src/main/java/datadog/trace/api/featureflag/exposure/Allocation.java]: **❔ question:** ‏Is that expected to redefine `equals()` without `hashCode()`?

File: internal-api/src/main/java/datadog/trace/api/featureflag/exposure/ExposureEvent.java [hunk 25]
```diff
@@ -0,0 +1,22 @@
+package datadog.trace.api.featureflag.exposure;
+
+public class ExposureEvent {
+  public final long timestamp; // System.currentTimeMillis()
```
Comment (PerfectSlayer) [inline internal-api/src/main/java/datadog/trace/api/featureflag/exposure/ExposureEvent.java]: **💭 thought:** `// System.currentTimeMillis()` is nice, I would add a comment about the unit too 😅 ‏

File: products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java [hunk 26]
```diff
@@ -0,0 +1,524 @@
...
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.regex.Pattern;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class DDEvaluator implements Evaluator, FeatureFlaggingGateway.ConfigListener {
```
Comment (PerfectSlayer) [inline products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java:42]: **❔ question:** Dumb question: ‏Should this class be part of lib too if it's loaded dynamically? 🤷

File: products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java [hunk 27]
```diff
@@ -0,0 +1,524 @@
...
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DDEvaluator.class);
+  private static final List<DateTimeFormatter> DATE_FORMATTERS =
+      asList(
+          DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").withZone(ZoneOffset.UTC),
+          DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'").withZone(ZoneOffset.UTC));
```
Comment (PerfectSlayer) [inline products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java]: **❔ question:** ‏Can the _optional section_ characters would work for this use case? (`[` and `]`).
I never used them to be honest but it seems that would serve this purpose: https://docs.oracle.com/en/java/javase/25/docs/api/java.base/java/time/format/DateTimeFormatter.html#patterns

File: products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java [hunk 28]
```diff
@@ -0,0 +1,524 @@
...
+      final EvaluationContext context) {
+    try {
+      final ServerConfiguration config = configuration.get();
+      if (config == null) {
+        return ProviderEvaluation.<T>builder()
+            .value(defaultValue)
+            .reason(Reason.ERROR.name())
+            .errorCode(ErrorCode.PROVIDER_NOT_READY)
+            .build();
+      }
```
Comment (PerfectSlayer) [inline products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java]: **🎯 suggestion:** ‏This would simplify the code to have an helper methods to return an evaluation error
like `error(T defaultValue, Reason reason)` and `error(T defaultValue, Reason reason, String message)`.

File: products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java [hunk 29]
```diff
@@ -0,0 +1,524 @@
...
+          .build();
+    }
+  }
+
+  private static boolean isEmpty(final List<?> list) {
+    return list == null || list.isEmpty();
+  }
+
+  private static boolean isAllocationActive(final Allocation allocation, final Date now) {
+    final Date startDate = parseDate(allocation.startAt);
```
Comment (PerfectSlayer) [inline products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java]: **❔ question:** ‏Not sure to have full context but would that help to store / cache date/time parsing into allocations if they get evaluated multiple times?

File: internal-api/src/main/java/datadog/trace/api/intake/Intake.java (hunk 21 above)
Comment (manuel-alvarez-alvarez) [reply internal-api/src/main/java/datadog/trace/api/intake/Intake.java:26]: Yeah, thought to do so, but it's an enum, I could convert it to a regular class if needed.

File: products/feature-flagging/agent/build.gradle.kts (hunk 22 above)
Comment (manuel-alvarez-alvarez) [reply products/feature-flagging/agent/build.gradle.kts]: `shadowJar` bundles the agent module (`FeatureFlaggingSystem`) with the lib module adding the RC service and also EVP proxy exposure writer in a single jar to be used in the build pipeline of the agent:
```
includeSubprojShadowJar(project(':products:feature-flagging:agent'), 'feature-flagging', includedJarFileTree)
```
Just removed the jar task as it's not really needed.

File: products/feature-flagging/api/build.gradle.kts (hunk 23 above)
Comment (manuel-alvarez-alvarez) [reply products/feature-flagging/api/build.gradle.kts:42]: No one depends on `:products:feature-flagging:api` module, this will be published to maven central for customers to use.

File: internal-api/src/main/java/datadog/trace/api/featureflag/exposure/Allocation.java (hunk 24 above)
Comment (manuel-alvarez-alvarez) [reply internal-api/src/main/java/datadog/trace/api/featureflag/exposure/Allocation.java]: No, actually this is a left over from a previous version of the exposure cache, sorry for that, just removed it.

File: .github/CODEOWNERS (hunk 20 above)
Comment (manuel-alvarez-alvarez) [reply .github/CODEOWNERS]: Done :smile:

File: internal-api/src/main/java/datadog/trace/api/featureflag/exposure/ExposureEvent.java (hunk 25 above)
Comment (manuel-alvarez-alvarez) [reply internal-api/src/main/java/datadog/trace/api/featureflag/exposure/ExposureEvent.java]: Done, thanks!

File: products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java (hunk 27 above)
Comment (manuel-alvarez-alvarez) [reply products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java]: Never use them myself, but they do work fine in this case, thanks for the input :smile:

File: products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java (hunk 29 above)
Comment (manuel-alvarez-alvarez) [reply products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java]: Yep, I moved it so it's done at parsing time.

File: products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java (hunk 26 above)
Comment (manuel-alvarez-alvarez) [reply products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java:42]: It's used by the provider which is loaded by the application classloader, meanwhile lib is loaded by the Datadog classloader.

File: products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java (hunk 28 above)
Comment (manuel-alvarez-alvarez) [reply products/feature-flagging/api/src/main/java/datadog/trace/api/openfeature/DDEvaluator.java]: Yep, done it, thanks!

File: internal-api/src/main/java/datadog/trace/api/intake/Intake.java (hunk 21 above)
Comment (PerfectSlayer) [reply internal-api/src/main/java/datadog/trace/api/intake/Intake.java:26]: Oh right, that's not worth it 😉

**Review by PerfectSlayer — APPROVED** [review]
Thanks for iterating with us about finding the right modular architecture 🙏 
Hopefully, we can built the standalone SDK easily from here.

File: dd-java-agent/agent-feature-flagging/build.gradle (hunk 5 above)
Comment (manuel-alvarez-alvarez) [reply dd-java-agent/agent-feature-flagging/build.gradle]: Changed, thanks!

File: dd-trace-api/src/main/java/datadog/trace/api/config/FeatureFlaggingConfig.java
Comment (jandro996) [inline dd-trace-api/src/main/java/datadog/trace/api/config/FeatureFlaggingConfig.java:1]: https://gitlab.ddbuild.io/DataDog/apm-reliability/dd-trace-java/-/jobs/1244579297

```
Execution failed for task ':dd-trace-api:jacocoTestCoverageVerification'.
> A failure occurred while executing org.gradle.internal.jacoco.JacocoCoverageAction
   > Rule violated for class datadog.trace.api.config.FeatureFlaggingConfig: instructions covered ratio is 0.0, but expected minimum is 0.9
```

Seems that this should be excluded  from jacocoTestCoverageVerification :)

Comment (manuel-alvarez-alvarez) [reply dd-trace-api/src/main/java/datadog/trace/api/config/FeatureFlaggingConfig.java:1]: Yep, fixed thanks!
